extensions/message-ids.md

```diff
@@ -0,0 +1,128 @@
+
+### Dependencies
+
+The `draft/message-tags` capability MUST be negotiated for servers wishing to use this tag.
+
+### Tags
+
+This specification adds the `draft/msgid` tag, which has one required value.
+
+This tag MAY be sent by a server on `PRIVMSG` and `NOTICE` messages.
+
+#### Tag value
+
+The tag value is chosen by the originating server and MUST be unique, meaning any other message transmitted on the entire network at any time MUST NOT share the same value.
+
+However, if a message is re-transmitted as-is, for example with the [`chathistory`](./batch/chathistory-3.3.html) batch type, the ID SHOULD be reused. As a result, clients MUST accept shared IDs.
+
+The tag value MUST be treated as a case sensitive opaque identifier. Clients MUST NOT use case folding or normalization when comparing IDs.
+
+## Server implementation considerations
+
+This section is non-normative.
+
+In order to guarantee sufficient uniqueness, message IDs can't be implemented as simple numeric counters that risk clashing with other servers on the network, or being reset if the server restarts.
+
+Some examples of appropriate IDs are:
+
+* An [RFC 4122 UUID](https://tools.ietf.org/html/rfc4122)
+* An ID generated by a shared external provider, used by all servers on the network
+* A numeric counter, prefixed with a server ID, with the latest value saved externally to the IRCd on every issue, to allow for power failures.
+* A numeric counter, prefixed with a server ID as well as the server startup timestamp at a suitable precision.
+
+When using timestamps, make sure they're correctly synchronised using NTP or similar.
+
+Although clients are required to treat IDs as case sensitive opaque values, servers might still choose a case insensitive ID format internally.
+
+### Choosing an ID format
+
+Certain message ID formats could present a length issue if other tags are being used. The trade-offs between length and complexity of implementation should be carefully considered when choosing an ID format.
+
+For instance, a UUID in hex form takes up 32 bytes (or 22 bytes in unpadded base64), and has widely available standardised implementations in many languages and no external dependencies. Conversely, a custom format can be much shorter, and might depend on an external service, disk I/O, additional state being tracked by the server, or untested proprietary code. 
+
+If the 512-byte tag limit is only being sparsely used overall, the simpler and longer format might be chosen. But if the server's capabilities put messages at a greater risk of reaching the limit, a shorter custom format might be more suitable.
+
+Another consideration is the chance of collisions. For example, some versions and implementations of the UUID specification rely on conditions that may not always produce collision-free outputs. However, if the probability is sufficiently low, a "practical" (if not guaranteed) uniqueness might be acceptable, given the requirement on clients to accept shared IDs.
```

Comment:
I think explicitly bringing this up just muddies the spec and isn't required. It bring up questions around exactly what a 'practical level of uniqueness' is, and weakens the "unique" language that's used through it.

imo we can just leave it as-is without this paragraph, say make it unique, and if a server really wants to do this and have a fantastically low chance of a collision happening then they'll decide whether it's an acceptable tradeoff.